src/main/java/at/ac/tuwien/kr/alpha/grounder/Substitution.java

```diff
@@ -68,29 +75,98 @@ public static Substitution unify(Literal literal, Instance instance, Substitutio
 	}
 
 	/**
-	 * Computes the unifier of the atom and the instance and stores it in the variable substitution.
-	 * 
-	 * @param atom         the body atom to unify
-	 * @param instance     the ground instance
-	 * @param substitution if the atom does not unify, this is left unchanged.
-	 * @return true if the atom and the instance unify. False otherwise
+	 * Helper class to lazily clone the input substitution of Substitution.unify only when needed.
 	 */
-	public static Substitution unify(Atom atom, Instance instance, Substitution substitution) {
-		for (int i = 0; i < instance.terms.size(); i++) {
-			if (instance.terms.get(i) == atom.getTerms().get(i) || substitution.unifyTerms(atom.getTerms().get(i), instance.terms.get(i))) {
-				continue;
+	private static class UnificationHelper {
+		Substitution returnSubstitution;
+
+		Substitution unify(List<Term> termList, Instance instance, Substitution partialSubstitution) {
+			for (int i = 0; i < termList.size(); i++) {
+				if (termList.get(i) == instance.terms.get(i) ||
+					unifyTerms(termList.get(i), instance.terms.get(i), partialSubstitution)) {
+					continue;
+				}
+				return null;
+			}
+			if (returnSubstitution == null) {
+				// All terms unify but there was no need to assign a new variable, return the input substitution.
+				return partialSubstitution;
 			}
-			return null;
+			return returnSubstitution;
 		}
-		return substitution;
+
+		boolean unifyTerms(Term termNonGround, Term termGround, Substitution partialSubstitution) {
```

Comment:
It seems wrong (or at least really confusing) to me that we now have `UnificationHelper::unifyTerms(Term, Term, Substitution)` as well as `Substitution::unifyTerms(Term, Term)`. The only difference seems to be that the one in `Substitution` always operates on `this` and modifying the substitution it's called on, while the other one operates on the given `partialSubstitution`. I think the version in `UnificationHelper` should be adapted so that the other one isn't needed.

Reply:
Agreed, the `Substitution::unifyTerms(Term, Term)` is only ever needed in unit tests, while the new one is not tested. Will adapt the tests to run on the real thing and get rid of the other.